[TH2-3070] Update codec core with required functionality for events reporting (#35)
th2-net/th2-codec: README.md

# Description (4.6.0)

This is a common codec library which takes care of some boilerplate stuff like subscribing/publishing to message queues, loading codec settings, etc.

# What is the codec?

The codec in th2 is a component that is responsible for transforming messages from human-readable format
into a format of the corresponding protocol and vise verse.
It contains the main logic for encoding messages for sending them to the system and decoding messages received from the system.

The codec communicates with other components by sending batches with groups of parsed or/and raw messages.

> During encoding, it transforms messages to a format of the corresponding protocol.

Reviewer: ```suggestion
During encoding, it transforms messages to the corresponding protocol format.
```

> During decoding, it takes all raw messages that correspond to the codec protocol and transform according to its rules.

Reviewer: ```suggestion
During decoding, it takes all raw messages that correspond to the codec protocol and transforms them according to its rules.
```

## Required pins

> Pins are the part of the main th2 concept. They describe what are the inputs and output of the box.

Reviewer: ```suggestion
Pins are a part of the main th2 concept. They describe what are the inputs and outputs of a box.
```

### Configuration example

This configuration is a general way for deploying components in th2.

> It contains the box configuration, pins' descriptions and other common parameters for a box.

Reviewer: ```suggestion
It contains box configuration, pins' descriptions and other common parameters for a box.
```

## Changelog

### v4.6.0

#### Fixed:

> * Errors and warnings during encoding does not have message IDs attached because they are not correct yet

Reviewer: > they are not correct yet

they - errors and warnings?

#### Added:

* Codec can report warnings during decoding and encoding message groups

#### Changed:

* Root codec event's name now uses box name
* The general encode/decode does not use `parentEventId` from messages when reporting errors and warnings
* The error/warning events are now attached to the root codec event.

> The error/warning event is attached to the event that is specified in `parentEventId` as reference to the event in codec root.

Reviewer: ```suggestion
* The error/warning event is attached to the event that is specified in `parentEventId` as a reference to an event in codec root.
```
